[#16] 리프레시 토큰 추가 + 쿠키 저장 설정

yohan-kang opened the pull request:
## Summary

>- close #16 

**리프레시 토큰 추가 + 쿠키 저장 설정 

## Tasks

- 로그인 요청시 엑세스토큰은 헤더로 ,리프레시 토큰은 쿠키로 응답
- 그에 따른 SecurityConfig 불필요 코드 제거 
- 로그인 관련 처리는 필터 처리로 로직이 구성되어 컨트롤러 단계에서 필요 없는 부분 코드를 주석 처리 했습니다 
[12/11] 재수정 
- api 모듈에서 jwt 처리하는 방향으로 개선
- 로그인 ,로그아웃, 재발급 로직 구현 + 레디스 연동 
- 코드 리뷰 내용 리펙토링 적용 
 
## To Reviewer

jwt 를 학습하면서 엑세스 토큰과 리프레시 토큰의 저장 경로를 다르게 두었습니다 

해결못한 이슈)api 모듈에서  /reissue 경로를  추가하여 리프레시 토큰을 사용하여 엑세스 토큰을 받아오는 코드 작업을 진행했지만 
2025-02-11T00:13:35.656+09:00 ERROR 30185 --- [team1] [nio-8080-exec-3] g.d.t.api.admin.application.AdminFacade  : Reissue failed: Message=Unable to make field private final byte[] java.lang.String.value accessible: module java.base does not "opens java.lang" to unnamed module @6e38921c, StackTrace={}
해당 오류의 원인을 찾지 못해 해당 부분은 후순위로 미루었습니다 


## Screenshot
로그인시 엑세스 토큰 발급 + 쿠키에 리프레시 토큰 응답 + 레디스에 리프레시 토큰 저장 확인  
<img width="1427" alt="스크린샷 2025-02-12 오후 3 22 56" src="https://github.com/user-attachments/assets/eceb3533-e7d1-458e-a1d1-7402798ac212" />
<img width="663" alt="스크린샷 2025-02-12 오후 3 27 41" src="https://github.com/user-attachments/assets/13684cd8-c58e-4693-874d-0c6f04361f33" />
<img width="825" alt="스크린샷 2025-02-12 오후 3 24 53" src="https://github.com/user-attachments/assets/53b18e17-129a-432d-bc57-86044bc70fcd" />



토큰 재발급 -> 재발급 요청시 쿠키를 통해 레디스에 리프레시토큰 갱신과 엑세스토큰,리프레시토큰 발급 확인  
<img width="1453" alt="스크린샷 2025-02-12 오전 11 35 14" src="https://github.com/user-attachments/assets/d0bdeecf-a577-4c91-ba01-8ee5f5bc7cff" />


로그아웃 시 쿠키,헤더초기화 확인 
<img width="1156" alt="스크린샷 2025-02-12 오전 11 35 57" src="https://github.com/user-attachments/assets/2e86b5c5-f47f-40ba-9849-96be556ac024" />

rmvmf96 committed:
feaf: SecurityCofig 수정 ,
LoginFilter (클라이언트의 로그인 요청을 처리하고, 인증이 성공하면 토큰을 발급) 추가
AdminAuthService (인증 요청이 들어왔을 때, DB에서 해당 사용자(Admin)를 조회하고 검증된 사용자 정보를 반환) 추가

rmvmf96 committed:
fix : 로그인 요청시 토큰발급을  컨트롤러에서 처리하지않고 Spring Security의 LoginFilter가 가로채서 처리하게 수정

rmvmf96 committed:
feat : 로그인 요청 성공,실패 요청 설정

rmvmf96 committed:
feat : 리프레시 토큰 생성 추가

rmvmf96 committed:
feat : 토큰 인증 로직 추가

rmvmf96 committed:
fix: 회원 가입 요청시 필터 검증 제외 하기

rmvmf96 committed:
fix : 로그인 요청시 잘못된 요청에서 RuntimeException 이 아닌  Spring Security 에서 처리 되어 response 가 출력되게 수정

rmvmf96 committed:
feat: 사용자의 권한 검증 설정 추가

rmvmf96 committed:
feat : 리프레시 토큰을 헤더에 담지 않고 쿠기에 저장으로 변경

rmvmf96 committed:
fix : 리프레시 토큰으로 엑세스 토큰을 새로 발급 받는 api 요청을 별로도 생성 예정 , 그에 따른 리프레시 토큰 검증 로직은 불필요 하므로 제거

minjo-on commented on team1-api/src/main/java/goorm/deepdive/team1/api/admin/application/AdminFacade.java:
```diff
@@ -14,17 +14,17 @@
         Admin admin = adminCommandService.register(request.email(), request.password(), request.role());
         String token = adminCommandService.login(request.email(), request.password());
         return new AdminRegisterResponse(admin.getId(), admin.getEmail(), token);
     }
```
회원 가입시엔 별다른 정보를 리턴할 필요가 없어 보입니다 다른 코드를 참고하면 좋을 것 같습니다

yohan-kang replied:
네 회원 가입 리턴값으로  가입한 스웨거에서는 응답으로  id 와 email만  확인 하도록 수정했습니다

minjo-on commented on team1-infra/src/main/java/goorm/deepdive/team1/infra/config/jwt/AdminAuthService.java:
```diff
@@ -0,0 +1,26 @@
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        Admin admin = adminRepository.findByEmail(email)
+                .orElseThrow(() -> new CustomException(AdminExceptionCode.ADMIN_NOT_FOUND));
+        // Admin 엔티티를 CustomAdminDetails로 감싸서 반환
```
에러 처리부분도 다른 부분을 참고하시면 좋을 것 같습니다

yohan-kang replied:
해당 패키지의 별도의 exception 을 생성하고 CustomException을 상속받은 클래스를 추가해서 수정해봤습니다 
혹시나 민준님이 생각하신 의도가 아니라면 말씀 남겨주시면 감사하겠습니다 
        Admin admin = adminRepository.findByEmail(email)
                .orElseThrow(AdminNotFoundException::new);

minjo-on commented on team1-infra/src/main/java/goorm/deepdive/team1/infra/config/jwt/CookieUtil.java:
```diff
@@ -0,0 +1,31 @@
+import jakarta.servlet.http.HttpServletResponse;
+
+public class CookieUtil {
+
```
쿠키는 CSRF 및 XSS 공격에 취약합니다. 그래서 이 대신에 캐시와 비슷한 성능을 가진 레디스를 사용한 것인데 쿠키를 사용한 이유가 있을까요?

yohan-kang replied:
로그인 요청시 사용자는 리프레시 토큰은 쿠키에 저장하고  서버는 레디스에 저장하는 방안으로 했는데 사용자에게 응답시 쿠키로 굳이 선택한 이유는 완벽한 보안은 아니지만 헤더보다는 안전하고 XSS 공격 경우  httpOnly를 설정으로 보안하고 CSRF 경우  Refresh 토큰을 사용하는 API는 토큰 재발급 API 단 하나이므로 CSRF의 공격 가치가 크지 않다고 생각해서 일단 쿠키에 저장해봤습니다

minjo-on commented on team1-infra/src/main/java/goorm/deepdive/team1/infra/config/jwt/JWTUtil.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Value("${spring.jwt.expiration-seconds}")
+    private long refresh_expirationSeconds;
+
```
따로 Properties 클래스를 만들어 환경변수를 관리하면 좋을 것 같습니다

yohan-kang replied:
네 해당 jwtUtil 파일에서 JwtProperties 클래스를 별도로 만들어서  jwtUtil 에서 환경변수를 관리하지 않게 수정했습니다

minjo-on commented on team1-infra/src/main/java/goorm/deepdive/team1/infra/config/jwt/JwtFilter.java:
```diff
@@ -0,0 +1,105 @@
+
+        // 액세스 토큰이 없거나 만료된 경우 처리
+        if (accessToken == null || jwtUtil.isExpired(accessToken)) {
+            log.warn("액세스 토큰이 만료되었거나 없습니다.");
```
검증하는 로직은 JwtUtill 클래스에 작성하는 이유가 있을까요? 필터에서 검증로직까지 직접 실행하는 것은 Srp 원칙에 위배된다고 생각합니다

yohan-kang replied:
네 해당부분 수정했습니다

minjo-on commented on team1-infra/src/main/java/goorm/deepdive/team1/infra/config/jwt/JwtFilter.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    protected boolean shouldNotFilter(HttpServletRequest request) throws ServletException {
+        String uri = request.getRequestURI();
+        return uri.contains("login") || uri.contains("swagger-ui") || uri.contains("swagger-resources")
```
이 부분은 SecurityConfig에서 해당 경로를 리스트로 정리해 permitAll로 설정하면 더 깔끔하게 코드 작성하여  해결할 수 있다고 생각되는데 따로 이 로직을 작성하신 이유가 있을까요?

yohan-kang replied:
permitAll()이 인증을 넘기지만 필터룰 건너뛰지는 않아서 해당 부분의 설정이 없으면 해당 경로로 요청을 하면  doFilterInternal 이 실행되어서 오류가 발생했습니다 그래서 Security의 필터 설정으로 해당 부분의 경로만 필터링하지 않도록 설정했습니다

minjo-on commented on team1-infra/src/main/java/goorm/deepdive/team1/infra/config/security/LoginFilter.java:
```diff
@@ -0,0 +1,114 @@
+
+
+public class LoginFilter extends UsernamePasswordAuthenticationFilter {
+
```
이렇게 구현하신 이유가 있을까요? 프로젝트가 Rest api를 사용하기 때문에 controller에서 직접 구현하면로그인 요청 → 서비스 호출 → 토큰 발급 → 응답이라는 직관적인 흐름을 유지할 수 있고, 단위 테스트와 통합 테스트 작성이 용이하기도 하면서, OIDC, 소셜 로그인, 2FA 등의 다양한 인증 방식 추가와 JWT 구조 및 리프레시 토큰 관리 등 커스터마이징이 쉬워지지 않을까 생각합니다

yohan-kang replied:
매번 로그인 로직경우 controller에서 구현을 했는데 이번 경우 로그인 요청에서 서비스 계층까지 가지 않고 필터에서 차단하면 낭비를 줄일수 있고 효율적이라고 생각을 하고 위협적인 악성 요청도 서비스 계층까지 도달하지 않으니 좋을수도 있다고 생각했습니다 

다만 민준님 답변처럼 직관적이고 테스트 용이성,확정성 면에서는 민준님 생각이 맞다고 저도 생각합니다 

일단 로그인 필터로 구현을 했지만 추 후 필요시에 개선을 해보겠습니다

minjo-on reviewed (commented):
작업하시느라 고생하셨습니다.  코멘트 확인해주시면 감사하겠습니다

minjo-on commented:
해당 오류는 Java 8 이상 버전에서 보안 강화로 인해 reflect API에 대한 접근이 제한되면서 발생하는 에러입니다. 이는 Java의 모듈 시스템(Jigsaw) 도입 이후, 특정 내부 API나 모듈에 대한 무단 접근을 방지하기 위해 기본적으로 접근 권한이 제한되었기 때문입니다.해결방법은 이를 접근 가능하기 위해 vm option에 
![image](https://github.com/user-attachments/assets/1bd23c63-ea61-4d46-ac6a-4d1c8b96fe1a)
이렇게 추가하시면 해결 될겁니다!

rmvmf96 committed:
fix : securityConfig 파일 위치 api 모듈로 위치를 변경
infra 모듈은 외부 시스템(redis,db) 담당 모듈이고 securityConfig 는 API 요청 처리(인증,인가) 이기 때문에 api 모둘에 배치되는게 맞다고 판단

rmvmf96 committed:
refactor : infra 모듈에 담겨있는 security, jwt 관련 로직 domain으로 변경

rmvmf96 committed:
refactor :  jwtFilter 검증 로직 수정 ,예외처리 파일추가

rmvmf96 committed:
refactor : infra 모듈의 있던 불필요 의존성 주입 파일 제거 ->  api 모듈로 이전

rmvmf96 committed:
refactor : 회원 가입 부분 예외처리 로직 수정 + 불필요 응답값 제거

rmvmf96 committed:
feat : 레디스 설정 추가

rmvmf96 committed:
feat : 로그인 요청시 레디스에 리프레시 토큰 저장 기능 추가

rmvmf96 committed:
feat : 토큰 재발급 api 요청을 위해 jwtUtil에 리프레시토큰 메서드 추가 , response dto 추가

rmvmf96 committed:
feat : 토큰 재발급 요청 api 추가

rmvmf96 committed:
feat : 로그아웃 기능 구현

rmvmf96 committed:
refacor : 관리자 api 스웨거 추가

rmvmf96 committed:
fix :  JwtUtil 잘못된 설정 오타 수정

rmvmf96 committed:
feat: 스웨거 로그인 api 추가

rmvmf96 committed:
refactor : 잘못 정의된 예외 처리 부분 수정 및 추가

rmvmf96 committed:
refactor : 사용하지 않은 메서드 제거

rmvmf96 committed:
refactor : 환경 변수 Properties 클래스로 처리하여 JwtUtil 클래스의  환경 변수 관리 책임 제거

rmvmf96 committed:
fix : jwt필터를 무시하기 위한 코드 추가

rmvmf96 committed:
Merge branch 'dev' into feature/#16-admin-jwt-Authentication

# Conflicts:
#	team1-api/src/main/java/goorm/deepdive/team1/api/admin/application/AdminFacade.java
#	team1-common/src/main/java/goorm/deepdive/team1/common/exception/AdminExceptionCode.java
#	team1-domain/build.gradle
#	team1-domain/src/main/java/goorm/deepdive/team1/domain/admin/application/AdminCommandService.java
#	team1-domain/src/main/java/goorm/deepdive/team1/domain/admin/application/AdminQueryService.java
#	team1-infra/src/main/java/goorm/deepdive/team1/infra/config/redis/RedisConfig.java
